DESIGN.md

```diff
@@ -0,0 +1,756 @@
+}
+```
+
+In practice, operation timeouts will deserve a dedicated API to handle the
+boilerplate.
+
+```rust
+let (tcp_stream, addr) = tcp_listener
+    .accept()
+    .timeout(Duration::from_millis(100))
+    .await?
+```
+
+The cancel and timeout function are inherent methods on operation future types.
+
+If the operation's future drops before the operation completes, the runtime will
+submit a cancellation request for the operation. However, cancellation is
+asynchronous and best-effort, the operation may still complete. In that case,
+the runtime discards the operation result.
+
+The queue's single-producer characteristic optimizes for a single thread to own
+a given submission queue. Supporting a multi-threaded runtime requires either
+synchronizing pushing onto the submission queue or creating a queue pair per
+thread. The tokio-uring runtime uses an isolated thread-per-core model,
+differing from Tokio's other runtimes. Unlike Tokio's primary multi-threaded
+runtime, there is no work-stealing. Tasks remain on the thread that spawned them
+for the duration of their lifecycle. Each runtime thread will own a dedicated
+submission and completion queue pair, and operations are submitted using the
+submission queue associated with the current thread. Operation completion
+futures will not implement `Send`, guaranteeing that they remain on the thread
+to receive the operation result. Interestingly, the resources, e.g.,
+`TcpListener`, can be `Send` as long as they do not hold operation futures
+internally. It is possible to have two operations in-flight from a single
+resource associated with different queues and threads.
+
+Because operations are not `Send`, tasks awaiting these operations are also not
+`Send`, making them unable to be spawned using the spawn function from the Tokio
+crate. The tokio-uring crate will provide a spawn function that accepts not
+`Send` tasks.
+
+When using multiple isolated runtime threads, balancing load between them
+becomes challenging. Applications must take care to ensure load remains balanced
+across threads, and strategies tend to vary. For example, a TCP server can
+distribute accepted connections across multiple threads, ideally while
+maintaining equal load across threads. One approach is to submit "accept"
+operations for the same listener concurrently across all the runtime threads
+while ensuring overloaded workers back-off. Defining runtime load is out of this
+crate's scope and left to the application, though pseudocode follows.
+
+```rust
+let listener = Arc::new(TcpListener::bind(...));
+
+spawn_on_each_thread(async {
+    loop {
+        // Wait for this worker to have capacity. This
+        // ensures there are a minimum number of workers
+        // in the runtime that are flagged as with capacity
+        // to avoid total starvation.
+        current_worker::wait_for_capacity().await;
+        
+        let socket = listener.accept().await;
+        
+        spawn(async move { ... });
```

Comment:
The `socket: TcpStream` is not itself a uring "operation", afaict, so it could be `Send`. Does that imply that load balancing could still happen between accepting connections and servicing them?

Load balancing of tasks to threads would then happen during `spawn`. You would `spawn` N tasks to `accept` (having all threads `accept` seems like overkill?), and upon `accept` they would `spawn` new tasks (which would likely go to other threads) to service the connections. The runtime would need to decide which thread to target each time it spawned a task, which could be tricky (long lived cheap tasks vs short lived expensive tasks, etc).

From an API perspective, the task would still be `!Send`, but the `spawn` function _would_ require `Send` for its argument(s) in order to move the inputs to the task onto its permanent home thread.

----

This question is particularly important I think because the effect of `current_worker::wait_for_capacity` seems to be that users end up creating their own thread pools (i.e. "a pool of threads fed by a queue") with the assistance of the `spawn_on_each_thread` function. In the accept case, the `accept` call is what waits on your queue: but other usecases that spawn work recursively would need an explicit queue of work after their `wait_for_capacity` call (if they actually wanted to balance across all threads that were available, rather than having recursion pinned at the top level).

It's not clear that any extra flexibility provided by this API justifies having applications need to implement their own load-balancing around `spawn` calls.

Reply:
Ah, sorry. Just noticed that this is addressed at the end in https://github.com/tokio-rs/tokio-uring/blob/design-doc/DESIGN.md#load-balancing-spawn-function . The idea that you might have load-balanced vs non-load-balanced spawn functions is interesting.